components/world_news_api/sources/news-event/news-event.mjs

```diff
@@ -0,0 +1,177 @@
+        "latestPublishedDate",
+      ],
+    },
+    newsSources: {
+      propDefinition: [
+        app,
+        "newsSources",
+      ],
+    },
+    authors: {
+      propDefinition: [
+        app,
+        "authors",
+      ],
+    },
+    entities: {
+      propDefinition: [
+        app,
+        "entities",
+      ],
+    },
+    locationFilter: {
+      propDefinition: [
+        app,
+        "locationFilter",
+      ],
+    },
+  },
+  methods: {
+    emit(data) {
+      this.$emit(data, {
+        id: data.id,
+        summary: data.title,
+        ts: Date.now(),
+      });
+    },
+    setLastId(id) {
+      this.db.set("lastId", id);
+    },
+    getLastId() {
+      return this.db.get("lastId");
+    },
+    async getCurrentPageRawData(offset, itemsPerPage, $) {
+      const params = {
+        "text": this.text || undefined,
+        "source-countries": getCommaSeparatedListFromArray(this.sourceCountries),
+        "language": this.language || undefined,
+        "min-sentiment": this.minSentiment || undefined,
+        "max-sentiment": this.maxSentiment || undefined,
+        "earliest-published-date": this.earliestPublishedDate || undefined,
+        "latest-published-date": this.latestPublishedDate || undefined,
+        "news-sources": getCommaSeparatedListFromArray(this.newsSources),
+        "authors": getCommaSeparatedListFromArray(this.authors),
+        "entities": getCommaSeparatedListFromArray(this.entities),
+        "location-filter": this.locationFilter || undefined,
+        "offset": offset,
+        "number": itemsPerPage,
+        "sort": "publish-time",
+        "sort-direction": "DESC",
+      };
+      const res = await this.app.searchNews(params, $);
+      return res;
+    },
+    getCurrentPageNewsArray(res, lastEmmitedId) {
+      const news = [];
+      let foundLastEmmitedId = false;
+      for (const news of res.news) {
+        if (news.id === lastEmmitedId) {
+          foundLastEmmitedId = true;
+          break;
+        }
+        news.unshift(news);
+      }
+
+      return {
+        news,
+        foundLastEmmitedId,
+      };
+    },
+  },
+  async run({ $ }) {
+    const ITEMS_PER_PAGE = 100;
+    const MAX_CALLS_PER_EXECUTION = 10;
+    const MAX_OFFSET = MAX_CALLS_PER_EXECUTION * ITEMS_PER_PAGE;
+
+    const lastEmmitedId = this.getLastId();
+    const newsToEmit = [];
+    let offset = 0;
+
+    do {
+      const res = await this.getCurrentPageRawData(offset, ITEMS_PER_PAGE, $);
+      if (res.news.length === 0) {
+        break;
+      }
+
+      const {
+        news,
+        foundLastEmmitedId,
+      } = this.getCurrentPageNewsArray(res, lastEmmitedId);
+      if (foundLastEmmitedId) {
+        break;
+      }
+
+      newsToEmit.push(...news);
+      offset += ITEMS_PER_PAGE;
+    } while (offset <= MAX_OFFSET);
+
+    if (newsToEmit.length > 0) {
+      this.setLastId(newsToEmit[0].id);
+    }
+
+    for (const news of newsToEmit.reverse()) {
+      this.emit(news);
+    }
```

Comment:
Regarding the whole logic of searching for news and emitting the events, I used the [Desk Check Methodology](https://www.techopedia.com/definition/18825/desk-check#:~:text=A%20desk%20check%20is%20an,the%20application%20from%20functioning%20properly.) to verify if it is indeed emitting the correct events and in the correct order. Please correct me if I'm wrong.

```js
line 148: const newsToEmit = [];
line 152: searches news in descending order by publish-time
              -> res.news = [most recent news, ... oldest]
line 160: for loop news in order, then unshifts
              -> news = [oldest news, ... most recent]
line 165: newsToEmit.push(...news)
              -> newsToEmit = [oldest news, ... most recent]
line 170: setLastId(oldest news id)
line 173: emits news reversed order
              -> emit = [most recent news, ... oldest news]
```

The logic above doesn't seem to be correct to me.

Here's a simplified pseudocode that would produce the correct results:
```js
search news in descending order by publish-time
    -> res.news = [most recent news, ... oldest]
loop search until last ID is found, then discard the rest
    -> newsToEmit.push(...res.news)
    -> newsToEmit = [most recent news, ... oldest]
setLastId(newsToEmit[0].id)
reversedNews = newsToEmit.reverse()
    -> reversedNews = [oldest news, ... most recent]
emit(news) for news in reversedNews
```

Reply:
I did one small change to fix one problem with the `lastId` check, but the idea behind the emit order still the same. Thanks